internal/keepers/util.go

```diff
@@ -198,6 +201,52 @@ func shuffledDedupedKeyList(attributed []types.AttributedObservation, key [16]by
 	return keys, nil
 }
 
+func assignMedianBlockToKeys(obs [][]ktypes.UpkeepKey) [][]ktypes.UpkeepKey {
+	var blockKeys []*big.Int
+	for _, ob := range obs {
+		for _, key := range ob {
+			blockKey, _, _ := key.BlockKeyAndUpkeepID()
+			blockKeyInt := big.NewInt(0)
+			blockKeyInt, _ = blockKeyInt.SetString(string(blockKey), 10)
+			blockKeys = append(blockKeys, blockKeyInt)
+		}
+	}
+
+	medianBlock := calculateMedianBlock(blockKeys)
+
+	var res [][]ktypes.UpkeepKey
+	for _, ob := range obs {
+		var ks []ktypes.UpkeepKey
+		for _, k := range ob {
+			_, upkeepID, _ := k.BlockKeyAndUpkeepID()
+			ks = append(ks, chain.UpkeepKey(fmt.Sprintf("%s|%s", medianBlock.String(), upkeepID)))
+		}
+		res = append(res, ks)
+	}
+
+	return res
+}
+
+func calculateMedianBlock(data []*big.Int) *big.Int {
+	dataCopy := make([]*big.Int, len(data))
+	copy(dataCopy, data)
+
+	sort.Slice(dataCopy, func(i, j int) bool {
+		return dataCopy[i].Cmp(dataCopy[j]) < 0
+	})
+
+	var median *big.Int
+	if l := len(dataCopy); l == 0 {
+		median = big.NewInt(0)
+	} else if l%2 == 0 {
```

Comment:
The correct way to calculate a median value across an even number of elements is to average the two middle elements, e.g. for a list of 1, 2, 3, 4, the median value would be (2 + 3) / 2 = 2.5. 

In this case, we'd have an invalid block number, but we could round the result up or down to give us a whole number/valid block number. 

The thing I'm not sure on is, should we only consider the set of block keys received as input, as candidates for the median value, or is it okay for the median value to not equal any of the input block keys?

E.g. for this list of block key: 10, 50, 100, 200, the median would be (50+100)/2 = 75. Is it okay for us to use 75 as the median block key, even though 75 wasn't passed in as part of the observations? Or should we instead limit ourselves to the set of input block keys only, e.g. in this case we'd use wither 50 or 100 as the median block key?

Reply:
I think in practice, the numbers should be close together and not spaced out a lot. In the case of an even number of values, I think it would be just fine if we selected the larger of the two. The primary goal is to eliminate extremes, not to get a perfect median. I would make a note in the code though on why we're not calculating a proper median. This could be confusing for later devs.